File: src/LTAdmin/Services/Business/EnrollmentService.cs
using LTAdmin.Core;
using LTAdmin.Data;
using LTAdmin.Models.Dto;
using LTAdmin.Models.Entities;
using LTAdmin.Repositories;
using LTAdmin.Services.Common;
using LTAdmin.Services.Logging;
using LTAdmin.Services.Referentiel;
using LTAdmin.Services.Validation;

namespace LTAdmin.Services.Business;

/// <summary>
/// Gestion des inscriptions : une ligne par étudiant et par classe
/// (historique + redoublements), contrôle de l'effectif maximum.
/// </summary>
public sealed class EnrollmentService : ServiceBase
{
    private readonly EnrollmentRepository _inscriptions;
    private readonly StudentRepository _etudiants;
    private readonly ReferentielRepository _referentiel;

    public EnrollmentService(AccessDatabase database, AppLogger logger, JournalService journal, ParametreService parametres)
        : base(database, logger, journal, parametres)
    {
        _inscriptions = new EnrollmentRepository(database);
        _etudiants = new StudentRepository(database);
        _referentiel = new ReferentielRepository(database);
    }

    public List<InscriptionDetail> ListByClasse(int idClasse)
    {
        try { return _inscriptions.ListByClasse(idClasse); }
        catch (Exception ex)
        {
            Logger.Error("Liste des inscrits d’une classe", ex);
            return new List<InscriptionDetail>();
        }
    }

    public List<InscriptionDetail> ListByEtudiant(int idEtudiant)
    {
        try { return _inscriptions.ListByEtudiant(idEtudiant); }
        catch (Exception ex)
        {
            Logger.Error("Historique des inscriptions d’un étudiant", ex);
            return new List<InscriptionDetail>();
        }
    }

    public InscriptionDetail? GetDetail(int idInscription)
    {
        try { return _inscriptions.GetDetail(idInscription); }
        catch (Exception ex)
        {
            Logger.Error("Lecture d’une inscription", ex);
            return null;
        }
    }

    public Inscription? Get(int idInscription)
    {
        try { return _inscriptions.GetById(idInscription); }
        catch (Exception ex)
        {
            Logger.Error("Lecture d’une inscription", ex);
            return null;
        }
    }

    /// <summary>Inscrit un étudiant dans une classe (contrôles : doublon, effectif max).</summary>
    public Result<int> Inscrire(Inscription inscription, string codeUtr)
    {
        var validation = EnrollmentValidator.Validate(inscription);
        if (!validation.IsValid) return Invalid<int>(validation);

        try
        {
            var etudiant = _etudiants.GetById(inscription.IdEtudiant!.Value);
            if (etudiant is null)
                return Result<int>.Fail("Étudiant introuvable.", ErrorCodes.NotFound);
            var classe = _referentiel.GetClasse(inscription.IdClasse!.Value);
            if (classe is null)
                return Result<int>.Fail("Classe introuvable.", ErrorCodes.NotFound);

            var doublon = _inscriptions.GetByEtudiantClasse(inscription.IdEtudiant.Value, inscription.IdClasse.Value);
            if (doublon is not null)
                return Result<int>.Fail($"« {etudiant.NomComplet} » est déjà inscrit dans la classe « {classe.Libelle} ».",
                    ErrorCodes.UniqueViolation);

            if (classe.EffectifMax.HasValue && classe.EffectifMax.Value > 0)
            {
                var effectif = _inscriptions.CountByClasse(classe.IdClasse!.Value);
                if (effectif >= classe.EffectifMax.Value)
                    return Result<int>.Fail(
                        $"Effectif maximum atteint pour « {classe.Libelle} » ({effectif}/{classe.EffectifMax}).",
                        ErrorCodes.BusinessRule);
            }

            if (string.IsNullOrWhiteSpace(inscription.NumInscription))
                inscription.NumInscription = GenererNumero();
            if (!inscription.DateInscription.HasValue)
                inscription.DateInscription = DateTime.Today;
            if (string.IsNullOrWhiteSpace(inscription.Statut))
                inscription.Statut = "INSCRIT";

            var id = _inscriptions.Insert(inscription);
            Journal.LogCreation(codeUtr, Tables.Inscription, id,
                $"{inscription.NumInscription} — {etudiant.NomComplet} → {classe.Libelle}");
            return Result<int>.Ok(id, $"Inscription {inscription.NumInscription} enregistrée.");
        }
        catch (Exception ex)
        {
            if (OleDbExceptionHelper.IsUniqueViolation(ex))
                return Result<int>.Fail("Cette inscription existe déjà (doublon étudiant × classe).",
                    ErrorCodes.UniqueViolation);
            return Failure<int>("Inscription d’un étudiant", ex);
        }
    }

    public Result Update(Inscription inscription, string codeUtr)
    {
        if (!inscription.IdInscription.HasValue)
            return Result.Fail("Inscription introuvable.", ErrorCodes.NotFound);
        var validation = EnrollmentValidator.Validate(inscription);
        if (!validation.IsValid) return Invalid(validation);

        try
        {
            var updated = _inscriptions.Update(inscription);
            if (updated == 0)
                return Result.Fail("Inscription introuvable : elle a peut-être été supprimée.", ErrorCodes.NotFound);
            Journal.LogModification(codeUtr, Tables.Inscription, inscription.IdInscription,
                inscription.NumInscription);
            return Result.Ok("Inscription enregistrée.");
        }
        catch (Exception ex)
        {
            if (OleDbExceptionHelper.IsUniqueViolation(ex))
                return Result.Fail("Cette inscription existe déjà (doublon étudiant × classe).",
                    ErrorCodes.UniqueViolation);
            return Failure("Modification d’une inscription", ex);
        }
    }

    /// <summary>Enregistre une sortie (date + motif) sans supprimer l'historique.</summary>
    public Result EnregistrerSortie(int idInscription, DateTime dateSortie, string? motif, string codeUtr)
    {
        try
        {
            var inscription = _inscriptions.GetById(idInscription);
            if (inscription is null)
                return Result.Fail("Inscription introuvable.", ErrorCodes.NotFound);
            inscription.DateSortie = dateSortie;
            inscription.MotifSortie = motif;
            inscription.Statut = "SORTI";
            var validation = EnrollmentValidator.Validate(inscription);
            if (!validation.IsValid) return Invalid(validation);
            _inscriptions.Update(inscription);
            Journal.LogModification(codeUtr, Tables.Inscription, idInscription, "Sortie enregistrée.");
            return Result.Ok("Sortie enregistrée.");
        }
        catch (Exception ex) { return Failure("Enregistrement d’une sortie", ex); }
    }

    public Result Delete(int idInscription, string codeUtr)
    {
        try
        {
            var existing = _inscriptions.GetById(idInscription);
            if (existing is null)
                return Result.Fail("Inscription introuvable.", ErrorCodes.NotFound);
            _inscriptions.Delete(idInscription);
            Journal.LogSuppression(codeUtr, Tables.Inscription, idInscription, existing.NumInscription);
            return Result.Ok("Inscription supprimée.");
        }
        catch (Exception ex)
        {
            if (OleDbExceptionHelper.IsForeignKeyViolation(ex))
                return Result.Fail("Suppression impossible : cette inscription possède des notes, absences, bulletins ou paiements.",
                    ErrorCodes.ForeignKeyViolation);
            return Failure("Suppression d’une inscription", ex);
        }
    }

    /// <summary>Génère un numéro d'inscription unique INS-AAAA-####.</summary>
    public string GenererNumero()
    {
        var prefixe = $"INS-{DateTime.Today.Year}-";
        var existants = new HashSet<string>(_inscriptions.ListNumeros(prefixe), StringComparer.OrdinalIgnoreCase);
        var sequence = 0;
        foreach (var numero in existants)
        {
            var suffixe = numero.Substring(prefixe.Length);
            if (int.TryParse(suffixe, out var value) && value > sequence)
                sequence = value;
        }
        string candidat;
        do
        {
            sequence++;
            candidat = prefixe + sequence.ToString("D4");
        } while (existants.Contains(candidat));
        return candidat;
    }
}

File: ltadmin/services/business/enrollment_service.py
"""Gestion des inscriptions : une ligne par étudiant et par classe
(historique + redoublements), contrôle de l'effectif maximum."""

from __future__ import annotations

import datetime as _dt
from typing import List, Optional

from ltadmin.core.result import Result, ResultValue
from ltadmin.data import error_helper
from ltadmin.data.access_database import AccessDatabase
from ltadmin.data.schema import Tables
from ltadmin.models.dto import InscriptionDetail
from ltadmin.models.entities import Inscription
from ltadmin.repositories.enrollment_repository import EnrollmentRepository
from ltadmin.repositories.referentiel_repository import ReferentielRepository
from ltadmin.repositories.student_repository import StudentRepository
from ltadmin.services.common import ServiceBase
from ltadmin.services.logging.app_logger import AppLogger
from ltadmin.services.logging.journal_service import JournalService
from ltadmin.services.referentiel.parametre_service import ParametreService
from ltadmin.services.validation.validation import EnrollmentValidator


class EnrollmentService(ServiceBase):

    def __init__(self, database: AccessDatabase, logger: AppLogger,
                 journal: JournalService, parametres: ParametreService):
        super().__init__(database, logger, journal, parametres)
        self._inscriptions = EnrollmentRepository(database)
        self._etudiants = StudentRepository(database)
        self._referentiel = ReferentielRepository(database)

    def list_by_classe(self, id_classe: int) -> List[InscriptionDetail]:
        try:
            return self._inscriptions.list_by_classe(id_classe)
        except Exception as ex:
            self.logger.error("Liste des inscrits d’une classe", ex)
            return []

    def list_by_etudiant(self, id_etudiant: int) -> List[InscriptionDetail]:
        try:
            return self._inscriptions.list_by_etudiant(id_etudiant)
        except Exception as ex:
            self.logger.error("Historique des inscriptions d’un étudiant", ex)
            return []

    def get_detail(self, id_inscription: int) -> Optional[InscriptionDetail]:
        try:
            return self._inscriptions.get_detail(id_inscription)
        except Exception as ex:
            self.logger.error("Lecture d’une inscription", ex)
            return None

    def get(self, id_inscription: int) -> Optional[Inscription]:
        try:
            return self._inscriptions.get_by_id(id_inscription)
        except Exception as ex:
            self.logger.error("Lecture d’une inscription", ex)
            return None

    def inscrire(self, inscription: Inscription, code_utr: str) -> ResultValue[int]:
        """Inscrit un étudiant dans une classe (contrôles : doublon, effectif max)."""
        validation = EnrollmentValidator.validate(inscription)
        if not validation.is_valid:
            return self.invalid_value(validation)
        try:
            etudiant = self._etudiants.get_by_id(inscription.id_etudiant)
            if etudiant is None:
                return ResultValue.fail("Étudiant introuvable.", "INTROUVABLE")
            classe = self._referentiel.get_classe(inscription.id_classe)
            if classe is None:
                return ResultValue.fail("Classe introuvable.", "INTROUVABLE")

            doublon = self._inscriptions.get_by_etudiant_classe(
                inscription.id_etudiant, inscription.id_classe)
            if doublon is not None:
                return ResultValue.fail(
                    f"« {etudiant.nom_complet} » est déjà inscrit dans la classe "
                    f"« {classe.libelle} ».", "DOUBLON")

            if classe.effectif_max is not None and classe.effectif_max > 0:
                effectif = self._inscriptions.count_by_classe(classe.id_classe)
                if effectif >= classe.effectif_max:
                    return ResultValue.fail(
                        f"Effectif maximum atteint pour « {classe.libelle} » "
                        f"({effectif}/{classe.effectif_max}).", "GESTION")

            if not (inscription.num_inscription or "").strip():
                inscription.num_inscription = self.generer_numero()
            if inscription.date_inscription is None:
                inscription.date_inscription = _dt.datetime.combine(
                    _dt.date.today(), _dt.time())
            if not (inscription.statut or "").strip():
                inscription.statut = "INSCRIT"

            inscription.id_inscription = self._inscriptions.insert(inscription)
            self.journal.log_creation(
                code_utr, Tables.INSCRIPTION, inscription.id_inscription,
                f"{inscription.num_inscription} — {etudiant.nom_complet} → {classe.libelle}")
            return ResultValue.ok(
                inscription.id_inscription,
                f"Inscription {inscription.num_inscription} enregistrée.")
        except Exception as ex:
            if error_helper.is_unique_violation(ex):
                return ResultValue.fail(
                    "Cette inscription existe déjà (doublon étudiant × classe).", "DOUBLON")
            return self.failure_value("Inscription d’un étudiant", ex)

    def update(self, inscription: Inscription, code_utr: str) -> Result:
        if inscription.id_inscription is None:
            return Result.fail("Inscription introuvable.", "INTROUVABLE")
        validation = EnrollmentValidator.validate(inscription)
        if not validation.is_valid:
            return self.invalid(validation)
        try:
            updated = self._inscriptions.update(inscription)
            if updated == 0:
                return Result.fail(
                    "Inscription introuvable : elle a peut-être été supprimée.", "INTROUVABLE")
            self.journal.log_modification(
                code_utr, Tables.INSCRIPTION, inscription.id_inscription,
                inscription.num_inscription)
            return Result.ok("Inscription enregistrée.")
        except Exception as ex:
            if error_helper.is_unique_violation(ex):
                return ResultValue.fail(
                    "Cette inscription existe déjà (doublon étudiant × classe).", "DOUBLON")
            return self.failure("Modification d’une inscription", ex)

    def enregistrer_sortie(self, id_inscription: int, date_sortie, motif: Optional[str],
                           code_utr: str) -> Result:
        """Enregistre une sortie (date + motif) sans supprimer l'historique."""
        try:
            inscription = self._inscriptions.get_by_id(id_inscription)
            if inscription is None:
                return Result.fail("Inscription introuvable.", "INTROUVABLE")
            inscription.date_sortie = date_sortie
            inscription.motif_sortie = motif
            inscription.statut = "SORTI"
            validation = EnrollmentValidator.validate(inscription)
            if not validation.is_valid:
                return self.invalid(validation)
            self._inscriptions.update(inscription)
            self.journal.log_modification(
                code_utr, Tables.INSCRIPTION, id_inscription, "Sortie enregistrée.")
            return Result.ok("Sortie enregistrée.")
        except Exception as ex:
            return self.failure("Enregistrement d’une sortie", ex)

    def delete(self, id_inscription: int, code_utr: str) -> Result:
        try:
            existing = self._inscriptions.get_by_id(id_inscription)
            if existing is None:
                return Result.fail("Inscription introuvable.", "INTROUVABLE")
            self._inscriptions.delete(id_inscription)
            self.journal.log_suppression(
                code_utr, Tables.INSCRIPTION, id_inscription, existing.num_inscription)
            return Result.ok("Inscription supprimée.")
        except Exception as ex:
            if error_helper.is_foreign_key_violation(ex):
                return Result.fail(
                    "Suppression impossible : cette inscription possède des notes, "
                    "absences, bulletins ou paiements.", "LIAISON")
            return self.failure("Suppression d’une inscription", ex)

    def generer_numero(self) -> str:
        """Génère un numéro d'inscription unique INS-AAAA-####."""
        prefixe = f"INS-{_dt.date.today().year}-"
        existants = {n.upper(): n for n in self._inscriptions.list_numeros(prefixe)}
        sequence = 0
        for numero in existants.values():
            suffixe = numero[len(prefixe):]
            try:
                value = int(suffixe)
            except ValueError:
                continue
            if value > sequence:
                sequence = value
        while True:
            sequence += 1
            candidat = prefixe + f"{sequence:04d}"
            if candidat.upper() not in existants:
                return candidat
